### PR TITLE
Fix/geo null

### PR DESCRIPTION
**Pull request by paulussup**
### Description

Add check for null and undefined.

#### Related issues
#1332
connect to https://github.com/strongloop/loopback-datasource-juggler/issues/1332

- None

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

File: lib/geo.js [hunk 1]
```diff
@@ -24,7 +24,7 @@ exports.nearFilter = function nearFilter(where) {
           var ret = nearSearch(el, parentKeys.concat(clauseKey).concat(index));
           if (ret) return ret;
         });
-      } else {
+      } else if (clause[clauseKey] !== null && clause[clauseKey] !== undefined) {
```
Comment (ebarault) [inline lib/geo.js]: we usually prefer early return to keep the code simpler to read, i suggest:

```
Object.keys(clause).forEach(function(clauseKey) {
  if (typeof clause[clauseKey] !== 'object' || !clause[clauseKey]) return;

  if (Array.isArray(clause[clauseKey])) {
    // array
  // object different than null, undefined, 0
  } else {
    // object other than null, undefined, 0
  }
}
```

File: test/basic-querying.test.js [hunk 2]
```diff
@@ -153,6 +153,27 @@ describe('basic-querying', function() {
           done();
         });
     });
+    it('should query by ids and null condition', function(done) {
```
Comment (ebarault) [inline test/basic-querying.test.js]: is it required for this test to search among these 4 users? unless i'm missing something could you simplify with only 2 different users?

File: test/basic-querying.test.js [hunk 3]
```diff
@@ -153,6 +153,18 @@ describe('basic-querying', function() {
...
+      User.findByIds([
+        createdUsers[0].id,
+        createdUsers[1].id],
+        {where: {vip: null}}, function(err, users) {
+          should.exist(users);
+          should.not.exist(err);
```
Comment (ssh24) [inline test/basic-querying.test.js]: @paulussup Could you please swap L161 with L162? It is always a good idea to check `err` first.

**Comment by ssh24** [comment]
@slnode test please

File: test/basic-querying.test.js (hunk 3 above)
Comment (paulussup) [inline test/basic-querying.test.js]: fixed @ssh24

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

File: test/basic-querying.test.js [hunk 4]
```diff
@@ -153,6 +153,18 @@ describe('basic-querying', function() {
           done();
         });
     });
+    it('should query by ids and null condition', function(done) {
+      User.findByIds([
+        createdUsers[0].id,
+        createdUsers[1].id],
+        {where: {vip: null}}, function(err, users) {
```
Comment (ssh24) [inline test/basic-querying.test.js]: @paulussup @ebarault I am not sure if this test case checks for the fix in the right away. The fix is however correct.

Shouldn't the filter do `near` query with a null value and see if the null value it is taken care of? Something like this test case but with a `null` value filter: https://github.com/strongloop/loopback-datasource-juggler/commit/64f64caaff70898281b09d8469db4e7a89fa160f#diff-8fae24fa21c82e23f7f43ccdae4fcf90R599

**Comment by paulussup** [comment]
To be honest I copy pasted the test above written by @raymondfeng and made the required edits. The reason that I chose to just test the null value is that the documentation uses querying a field as null as an example. Because this was not tested the error was able to make its way in to master.

**Comment by ssh24** [comment]
@paulussup That is fine. We can still keep this test case and just add that extra test case.

The test case is also failing on mongodb and cassandra connector so need to look into that.

**Comment by paulussup** [comment]
I guess that is because those datasources don’t default the value to null when persisting like the sql datasources.

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@ebarault This fix looks good to me. Could you do me a favour and review the changes I committed, i.e fixing and adding new test cases?

@slnode test please

**Comment by ssh24** [comment]
@slnode test please

File: test/basic-querying.test.js [hunk 5]
```diff
@@ -153,6 +153,26 @@ describe('basic-querying', function() {
           done();
         });
     });
+    it('should query by ids and null condition', function(done) {
+      // due to some connectors not being assigning null value by default
+      // the query is filtered by the fields which may consist null value
```
Comment (ebarault) [inline test/basic-querying.test.js]: typo: "which may consist **in** null value ..."

File: test/basic-querying.test.js [hunk 6]
```diff
@@ -153,6 +153,26 @@ describe('basic-querying', function() {
...
+      // or not exist at all
+      User.findByIds([
+        createdUsers[0].id,
+        createdUsers[1].id],
+        {fields: {name: true, vip: true}}, function(err, users) {
```
Comment (ebarault) [inline test/basic-querying.test.js]: i'm afraid i don't understand, the test is named: 'should query by ids and null condition' and we are not querying by null value here. I suspect it should be `vip: null` here?

File: test/basic-querying.test.js [hunk 7]
```diff
@@ -593,6 +613,19 @@ describe('basic-querying', function() {
 
     describe('geo queries', function() {
       describe('near filter', function() {
+        it('supports a near query with null condition', function(done) {
```
Comment (ebarault) [inline test/basic-querying.test.js]: could you move this test right after the 'supports a basic "near" query' test please?

File: test/basic-querying.test.js (hunk 6 above)
Comment (ebarault) [inline test/basic-querying.test.js]: or i'm missing this: one of the user has an undefined property "vip" and the test should be named something like: `'should query by ids and condition, allowing results where conditions match null properties'`

File: test/basic-querying.test.js (hunk 7 above)
Comment (ebarault) [inline test/basic-querying.test.js]: if i'm right in my previous comment the test name should also be revised to reflect that we are not querying on a null condition, but that results on the conditions could come with empty properties

something like:
`'supports a compound near query with conditions, allowing results where conditions match null properties'`

**Comment by ssh24** [comment]
@ebarault Updated the test messages.

@slnode test please

File: test/basic-querying.test.js (hunk 6 above)
Comment (paulussup) [inline test/basic-querying.test.js]: I have to say I don't see how this test is even testing the behaviour that caused this bug. The problem arose when traversing the where object for the near key and the value is``` {where:
 {field: null}}```. Would it not make more sense to set vip explicitly to null in this test case or create new test seed data with explicit null values.

Comment (ssh24) [inline test/basic-querying.test.js]: @paulussup Good question. Your test case was not wrong however it had a loophole where it would not check for connectors which does not put `null` values for fields which are not set (it would instead have it as undefined, for example mongodb).

Improved the test case to take care of that situation by filtering by fields (name and vip) and then get those instances back and then perform a check if they are either `null` or `undefined`. After fixing this it works with the downstream connectors such as mongodb and cassandra.

Comment (paulussup) [inline test/basic-querying.test.js]: I accept that it fixes the downstream issue. If there was a test which actually set the value explicitly to null in the test I think this too would allow the downstream connectors to query the null value and return records. Mongo and Cassandra can both store null values.

Comment (ssh24) [inline test/basic-querying.test.js]: @paulussup Yes I agree with you. Would you like to fix the test cases to go with the changes you proposed? We can just add a property `vip` with value **null** on the second user of `createdUsers` and just query that with your previous test case.

Please be sure to pull latest from your branch if you haven't already.

**Comment by paulussup** [comment]
Sure no problems.

**Comment by paulussup** [comment]
Haven’t had a chance yet.

**Comment by ssh24** [comment]
@paulussup I have fixed the test cases myself.

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
PR ready to be landed. @paulussup Thank you again for the patch!
